hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -8,9 +8,399 @@ import is.hail.expr.types.physical._
+  def toEmitTriplet(opt: COption[Code[_]], ti: TypeInfo[_], mb: MethodBuilder): EmitTriplet = {
+    val m = mb.newLocal[Boolean]
+    val v = mb.newLocal(ti)
+    val setup = JoinPoint.CallCC[Unit] { (jb, ret) =>
+      opt(Code(m := true, v.storeAny(defaultValue(ti)), ret(())),
+          a => Code(m := false, v.storeAny(a), ret(())))(EmitStreamContext(mb, jb))
+    }
+    EmitTriplet(setup, m, v.load())
+  }
+  def toTypedTriplet[A](t: PType, mb: MethodBuilder, opt: COption[Code[A]]): TypedTriplet[t.type] =
+    TypedTriplet(t, toEmitTriplet(opt, typeToTypeInfo(t), mb))
+}
+
+object CodeStream { self =>
+  import is.hail.asm4s.joinpoint.JoinPoint.CallCC
+  import is.hail.asm4s.joinpoint._
+  def newLocal[T: ParameterPack](implicit ctx: EmitStreamContext): ParameterStore[T] = implicitly[ParameterPack[T]].newLocals(ctx.mb)
+  def joinPoint()(implicit ctx: EmitStreamContext): DefinableJoinPoint[Unit] = ctx.jb.joinPoint()
+  def joinPoint[T: ParameterPack](implicit ctx: EmitStreamContext): DefinableJoinPoint[T] = ctx.jb.joinPoint[T](ctx.mb)
+
+  private case class Source[+A](setup0: Code[Unit], close0: Code[Unit], setup: Code[Unit], close: Code[Unit], pull: Code[Ctrl])
+
+  abstract class Stream[+A] {
+    private[CodeStream] def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A]
+
+    def fold[S: ParameterPack](s0: S, f: (A, S) => S, ret: S => Code[Ctrl])(implicit ctx: EmitStreamContext): Code[Ctrl] =
+      CodeStream.fold(this, s0, f, ret)
+    def forEach(f: A => Code[Unit], ret: Code[Ctrl])(implicit ctx: EmitStreamContext): Code[Ctrl] =
+      CodeStream.forEach(this, f, ret)
+    def forEach(mb: MethodBuilder)(f: A => Code[Unit]): Code[Unit] =
+      CallCC[Unit]((jb, ret) => CodeStream.forEach(this, f, ret(()))(EmitStreamContext(mb, jb)))
+    def mapCPS[B](
+      f: (EmitStreamContext, A, B => Code[Ctrl]) => Code[Ctrl],
+      setup0: Option[Code[Unit]] = None,
+      setup:  Option[Code[Unit]] = None,
+      close0: Option[Code[Unit]] = None,
+      close:  Option[Code[Unit]] = None
+    ): Stream[B] = CodeStream.mapCPS(this)(f, setup0, setup, close0, close)
+    def map[B](
+      f: A => B,
+      setup0: Option[Code[Unit]] = None,
+      setup:  Option[Code[Unit]] = None,
+      close0: Option[Code[Unit]] = None,
+      close:  Option[Code[Unit]] = None
+    ): Stream[B] = CodeStream.map(this)(f, setup0, setup, close0, close)
+    def flatMap[B](f: A => Stream[B]): Stream[B] =
+      CodeStream.flatMap(map(f))
+  }
+
+  implicit class StreamPP[A](val stream: Stream[A]) extends AnyVal {
+    def filter(cond: A => Code[Boolean])(implicit pp: ParameterPack[A]): Stream[A] =
+      CodeStream.filter(stream, cond)
+  }
+
+  def unfold[A, S: ParameterPack](
+    s0: S,
+    f: (S, EmitStreamContext, COption[(A, S)] => Code[Ctrl]) => Code[Ctrl]
+  ): Stream[A] = new Stream[A] {
+   def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A] = {
```

Comment:
not worth changing in this PR, but I have a preference for explicit parameters rather than implicit ones, both for readability and better IDE support.

Reply:
It also seems like in many cases below you use `ctx.mb.newLocal` (which is very clear), so this should be an easy change to make afterwards.

Reply:
A made a note to go back and change that later.